example/kernelSpecialization/src/kernelSpecialization.cpp

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2021 Sergei Bastrakov
+ *
+ * This file exemplifies usage of alpaka.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS” AND ISC DISCLAIMS ALL WARRANTIES WITH
+ * REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+
+#include <iostream>
+
+//#############################################################################
+//! Kernel to illustrate specialization for a particular accelerator
+//!
+//! It has a general operator() implementation and a specialized one for the CUDA accelerator.
+//! When running the kernel on a CUDA device, the specialized version of operator() is called.
+//! Otherwise the general version is called.
+//! The same technique can be applied for any function called from inside the kernel,
+//! thus allowing specialization of only relevant part of the code.
+//! It allows accelerator-specific optimization or accessing specific functionality not abstracted by alpaka.
+struct Kernel
+{
+    //-----------------------------------------------------------------------------
+    //! Implementation for the general case
+    //!
+    //! It will be called when no specialization is a better fit.
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc) const
+    {
+        // For simplicity assume 1d thread indexing
+        auto const globalThreadIdx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0];
+        if(!globalThreadIdx)
```

Comment:
Can you make it more explicit, that thread 0 is running the printf function? At the first time, when I read the code, I thought the if statement do some error checking because of the `!` operator.

Reply:
Agree, will change.

Reply:
Fixed.